### DISCUSSION TITLE
Solis Hybrid "Allow Grid Charging"

### DISCUSSION THREAD

**Discussion by mart-mac** [wills106/homeassistant-solax-modbus#410] [Q&A]
I'm trying to find if this is any option already, and if not, weather it can be added

Within the "self use" menu there is another option to enable / disable grid charging, Can this be changed with this integration currently?

Thanks

**Comment by alienatedsec** [comment]
@mart-mac check here https://github.com/wills106/homeassistant-solax-modbus/issues/405#issuecomment-1549826641

**Reply by mart-mac** [reply]
#405 is just about turning the charge window on and off,  I'm happy and got automations working with that

I'm currently trying to limit my charge amps during sunny days so my battery doesn't fill up so quickly by setting a charge window with limited amps.  If my battery get full early in the day I'm limited to my 6kw out put,  keep the battery slow charging I'm hitting 7-8kw when the suns out.  The problem I've had some times is when big clouds comes over and I've set it to charge at 2kw, if my PV drop below 2kw it draws from the grid to charge.  If I disable grid charge (on the inverter) I don't have this problem, but as soon as you activate storage mode "Times charge/discharge" on the integration it switched grid charge back on, with no way to turn back off

**Reply by alienatedsec** [reply]
Changing to "Auto Mode" is switching the grid charge off.

**Comment by Cabocab** [comment]
Grid charging is enabled by bit 5 of the energy storage control switch register. By default, the integration sets the grid charging to on. I was able to set grid charging to off while in timed charge/discharge mode by writing 3 instead of 35 to the register.  This retains the functionality of the timed charge/discharge mode but limits the battery charging to solar only.

**Reply by wills106** [reply]
I can add in more modes if you can provide mode names and what to write etc

**Reply by mart-mac** [reply]
> Grid charging is enabled by bit 5 of the energy storage control switch register. By default, the integration sets the grid charging to on. I was able to set grid charging to off while in timed charge/discharge mode by writing 3 instead of 35 to the register. This retains the functionality of the timed charge/discharge mode but limits the battery charging to solar only.

This is what I'm trying to do.  

@wills106 I think you'd call this mode timed charge/discharge with no grid charge.  As on the inverter it is a sub menu

**Reply by Cabocab** [reply]
@wills106 I think that the grid charge flag would only be relevant to auto mode and timed charge discharge mode, which would also be consistent with the inverter gui (and these are the only modes that I have tested so far). 

Auto mode with no grid charge : 1
Timed charge/discharge with no grid charge : 3

**Reply by alienatedsec** [reply]
@Cabocab have you read the register after changing the mode on the inverter? If not, let me know what changes you made and I can read exactly what is set on the register.

**Reply by Cabocab** [reply]
@alienatedsec yes, I read the registers (33132 read only and 43110 holding register) from the inverter with both grid charging on and off to verify which bit controls grid charging. 

Leading zeros omitted for brevity. 
For auto mode with grid charging register is 100001
For auto mode with no grid charging register is 1
For timed use with grid charging register is 100011
For timed use with no grid charging register is 11.

I wasn’t able to try the other modes as they weren't relevant to my use case (off grid and battery wake up modes) but I could try them out when I have some free time. 

I added the 2 new modes (1 and 3) to option_dict for both 33132 and 43110 registers in plugin_solis.py

To clarify, when I say bit 5 (32), I start counting from bit 0 (1).

**Reply by wills106** [reply]
I'm just adding these in:

> I was able to set grid charging to off while in timed charge/discharge mode by writing 3 instead of 35 to the register.

Is this different to 

> Timed charge/discharge with no grid charge : 32

Might be easier if you just paste in your 33132 and 43110 registers then I can add them into the Integration

**Reply by wills106** [reply]
This is what I have:
```
SolisModbusSensorEntityDescription(
        name = "Energy Storage Control Switch",
        key = "energy_storage_control_switch",
        register = 33132,
        register_type = REG_INPUT,
        scale = {  
                1: "Auto Mode - No Grid Charge",
                3: "Timed Charge/Discharge - No Grid Charge",
                33: "Auto Mode",
                35: "Timed Charge/Discharge", 
                37: "Off-Grid Mode", 
                41: "Battery Awaken", 
                43: "Battery Awaken + Timed Charge/Discharge",
                51: "Backup/Reserve",
                },
        allowedtypes = HYBRID,
    ),
```

Is that correct?

Edit or should '3' be 32: "Timed Charge/Discharge - No Grid Charge",

**Reply by Cabocab** [reply]
This is correct.  Apologies for the confusion.  That was a typo in my earlier reply.  Should be '3' instead of '32'.  Have edited my earlier reply.

> ```
> SolisModbusSensorEntityDescription(
>         name = "Energy Storage Control Switch",
>         key = "energy_storage_control_switch",
>         register = 33132,
>         register_type = REG_INPUT,
>         scale = {  
>                 1: "Auto Mode - No Grid Charge",
>                 3: "Timed Charge/Discharge - No Grid Charge",
>                 33: "Auto Mode",
>                 35: "Timed Charge/Discharge", 
>                 37: "Off-Grid Mode", 
>                 41: "Battery Awaken", 
>                 43: "Battery Awaken + Timed Charge/Discharge",
>                 51: "Backup/Reserve",
>                 },
>         allowedtypes = HYBRID,
>     ),
> ```

**Reply by wills106** [reply]
I thought that was correct and what I released in the latest Beta yesterday.

**Reply by mart-mac** [reply]
I've just update to your latest Beta, and this is exactly what I wanted. 

Thanks all
